### ISSUE TITLE
could not track when padding occur in wmt problems

### ISSUE THREAD

**Issue by colmantse** [tensorflow/tensor2tensor#221]
Dear all,

It is not clear to me that padding has occurred at the character_generator in WMT.py. I further trace back to text_encoder.bytetextencoder() but it doesnt seems to be there either. For a dynamic sequence to sequence problem, padding must have had occurred somewhere, may someone show me where to look?

Best,
Colman

**Comment by lukaszkaiser** [comment]
Hi @colmantse , thanks for asking. We don't do padding in t2t-datagen, data is saved example-by-example without any padding. Then, when data is read in `data_reader.py`, we bucket it by sequence length and add padding when constructing batches using `tf.contrib.training.bucket_by_sequence_length` here:
https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/data_reader.py#L283

Hope that helps! We're preparing a doc on the whole journey of an example in T2T that will hopefully make things easier to find, it can be quite hard at present. I'm closing this, but don't hesitate to re-open or make new issues or ask questions on the mailing list!

**Comment by colmantse** [comment]
Thanks much! I really appreciate it, you guys are awesome. I was looking for it the whole afternoon yesterday, never for a second suspect it would be data-reader.py lol

**Comment by colmantse** [comment]
Hi @lukaszkaiser ,

I have been reading the codes, but I could not quite detect the stage where the input is transformed from tf.VarLenFeature(tf.int64) to embedding. I read the paper and it seems a matching of token_id with embedding does take place in both input and output. Would be great to know where to look.

Best,
Colman
